### PR TITLE
Import and command line interfaces + lots of fixes for issues

### PR DESCRIPTION
**Pull request by MathInspector**
Before merging these changes into master, I'd like to give everyone a chance to test everything, including myself.

Besides fixing a large number of issues, this branch also has both the import and command line interfaces.

### CLI
```
usage: mathinspector [files] ... [options] ... 

Options and arguments:
[files]...  : a list of files with either a .math or .py extension.  The .math
                   file will be loaded, and all of the .py files will be added to
                   the current project
--help obj  : view the documentation for obj as if you called help(obj) in
                    mathinspector
--new       : starts a new project and resets the state of the app.  This flag will
             overwrites the autosave file with a blank file
```

### Import
```python
>>> import mathinspector

>>> mathinspector.viewdoc(mathinspector)

>>> mathinspector.main()

>>> mathinspector.plot(1,2,3)
```

both of these respect the existing autosave file when you launch main with .math and/or .py files, and all three interfaces (cli, import, app) maintain a strong amount of symmetry with how everything is accessed in the mathinspector source code files themselves.

The following issues have been fixed in this branch:
https://github.com/MathInspector/MathInspector/issues/24
https://github.com/MathInspector/MathInspector/issues/26
https://github.com/MathInspector/MathInspector/issues/27
https://github.com/MathInspector/MathInspector/issues/28
https://github.com/MathInspector/MathInspector/issues/29
https://github.com/MathInspector/MathInspector/issues/30
https://github.com/MathInspector/MathInspector/issues/32
https://github.com/MathInspector/MathInspector/issues/34
https://github.com/MathInspector/MathInspector/issues/35
https://github.com/MathInspector/MathInspector/issues/36

While fixing https://github.com/MathInspector/MathInspector/issues/29 I discovered a number of other bugs associated with this setup, and am planning to work on that, which will help verify that these changes are all working properly.

### CODE REVIEW

File: .gitignore
Comment (MathInspector) [reply .gitignore:5]: I have them in my local for testing on linux at the moment, feel free to remove that if you think its better.  I was trying to completely remove vagrant from the python package to be more like other packages

File: mathinspector/console/interpreter.py
Comment (MathInspector) [reply mathinspector/console/interpreter.py:281]: whoops!

**Comment by MathInspector** [comment]
> the __init__/__main__/app/*.spec stuff is definitely broken, and the version handling changes too.

I have a plan for pyinstaller which will require changes, good call.  As for the version handling changes, I confess I do not completely understand them yet and am trying to figure it out as I go.  I can't seem to get the version to show correctly when running from source without using a string.

File: mathinspector/console/interpreter.py
Comment (casperdcl) [inline mathinspector/console/interpreter.py:281]: this isn't at all like `sys.stdout.buffer` though...

Comment (MathInspector) [reply mathinspector/console/interpreter.py:281]: I tried to read up on this and then did my best to guess how it works based on my understanding of the word "buffer" and how I thought it worked.

Which part is wrong?

**Comment by casperdcl** [comment]
> when running from source

If you're running from source you need to manually install some dependencies :)

https://github.com/pypa/pip/issues/8049#issuecomment-785401708

File: mathinspector/console/interpreter.py
Comment (casperdcl) [reply mathinspector/console/interpreter.py:281]: well first up `sys.stdout` inherits from `TextIOWrapper`. There be a bushel of ye moste dark magike in mathinspector overloading builtin functionality but you can read up on io buffers here: https://docs.python.org/3/library/io.html

Comment (MathInspector) [reply mathinspector/console/interpreter.py:281]: new idea
```python
sys.stdout = StdOut(sys.stdin, self.write)
sys.stderr = StdOut(sys.stdin, self.write)

...

class StdOut(TextIOWrapper):
    def __init__(self, buffer, write, **kwargs):
        super(StdOut2, self).__init__(buffer, **kwargs)
```

It seems to work.  Before pushing this, I am wondering if you like this better or if it needs more work?
Screen Shot 2021-02-25 at 12 34 14 PM
        self.write = write

**Comment by MathInspector** [comment]
@casperdcl would you like to be included in the credits in the app?
Screen Shot 2021-02-25 at 1 22 44 PM

File: mathinspector/console/interpreter.py
Comment (casperdcl) [reply mathinspector/console/interpreter.py:281]: `sys.stdin` are you sure? :)

File: mathinspector/util/common.py
Comment (MathInspector) [reply mathinspector/util/common.py]: os.system has been completely removed everywhere and replaced with subprocess.run in the latest push to the dev branch

File: .gitignore
Comment (MathInspector) [reply .gitignore]: All vagrant stuff is out of .gitignore now.  I almost kept install.sh but figured it was better to remove that file

I created a new repository for all vagrant related items.  I have some plans for this repo, but at the moment it's main function is to say it's where I am planning to put tests at an unspecified time in the future (in case anyone asks if there are tests)

https://github.com/MathInspector/mathinspector-vagrant

File: mathinspector/console/interpreter.py
Comment (MathInspector) [reply mathinspector/console/interpreter.py]: > sys.stdin are you sure? :)

lol, I am not sure about this at all, except in as much as it seems to work.  Some of the previous black magic didn't seem required in interpreter.py any longer, and I checked around to try and break stuff, it seems fine.  This is even darker black magic and I like it more :)

**Comment by MathInspector** [comment]
Thanks so much for taking the time to review this PR.  Let me know if your requested changes have been addressed or not, or if you have more requests after the latest push.  I think it's good to go now, however, I am going to continue working on things and if i find any bugs will push them to the dev branch.

> If you're running from source you need to manually install some dependencies

I just created a file _dist_version.py and put a comment next to it.

By the way, version finally moved to mathinspector/config.py https://github.com/MathInspector/MathInspector/blob/dev/mathinspector/config.py

That is where I silo'd off all the platform specific configuration stuff from mathinspector/util

**Comment by MathInspector** [comment]
I restored install.sh because of the issue related to copying files into ttkthemes.  My understanding is the way the package is built requires that the file copying into ttkthemes has been done already, and I changed my mind about removing install.sh after thinking it's probably best to document the ttkthemes stuff with working code.

**Comment by MathInspector** [comment]
The latest push has the fix for https://github.com/MathInspector/MathInspector/issues/37 and the entire main menu layout has been redone with the suggestions from @EtienneDesticourt

**Comment by casperdcl** [comment]
> would you like to be included in the credits in the app?

Well I've just added our project to https://www.caspersci.uk.to/work/#foss which means a few extra hits per day :)

**Comment by casperdcl** [comment]
ok just forced-pushed after rebase c853047...f695824 (no difference, just keeps history cleaner). Will have a look to see if anything's broken :)

**Comment by MathInspector** [comment]
I am a very prolific committer.  Also, I wish I was at good at spotting bugs *before* pushing code to remote as I am after =P

I feel a little embarrassed about the doc's style suffering a regression where some of the parsing that used to works is broken now.  I don't want to bloat this PR with a large number of additional commits, and might even start working on a new branch for the sake of getting this PR wrapped up.

I also don't want to create a precedent for things not being perfect, but these are the tradeoffs that one must balance while working on a new project.

**Comment by casperdcl** [comment]
I don't have a strong opinion on "delaying" but I do think in general closing 10+ issues with one PR is excessive. Nearly done reviewing so far.

**Review by casperdcl — CHANGES_REQUESTED** [review]
1. The `sys.modules[__name__] = mathinspector()` is a super hacky way to prevent importing other things... The normal way to do this is to move `__init__` to `app` and only import things in `__init__` which you want exposed. Logically it's also all app-specific logic so doesn't belong in `__init__` (e.g. someone calling `mathinspector --help` on the CLI does not need the app to run at all).

Is there another reason why you do this?

I'm going to close my eyes over the other dark magicke of redefining `help()`, `print()`, `sanity()`, etc...

2. the `StdWrap` stuff still doesn't seem to handle `flush()` correctly.

```python
import tqdm, time
# doesn't print anything for 2sec in mathinspector REPL
for i in tqdm.trange(10):
    time.sleep(0.2)
```

3. I think leaving an **Animation** heading in the readme somewhere is a good idea.

I've fixed:

a) `__version__`
b) PyIntstaller (without needing `_run.py` in the `mathinspector/` directory)
c) problem when right-clicking nodes
d) potential problem with `StdWrap.encoding`

File: mathinspector/__init__.py
Comment (casperdcl) [inline mathinspector/__init__.py:112]: zoïks

File: mathinspector/__init__.py
Comment (casperdcl) [inline mathinspector/__init__.py:216]: why in a different place from `stdout` wrapping?

**Comment by MathInspector** [comment]
> I don't have a strong opinion on "delaying" but I do think in general closing 10+ issues with one PR is excessive

I decided I don't want to push this to master with the doc stuff like it is.  I'm going to work on that and address the things you are raising.

It can be merged into master when it's ready.

File: mathinspector/__init__.py
Comment (MathInspector) [reply mathinspector/__init__.py:216]: When sys.stderr is in Interpreter, and there is a syntax error in the math inspector source code, the app will fail to launch and there are no error messages.

So I moved it into `main` to be sure to be able to see error messages when that happens, and left the other stuff where it is while trying to figure out if all override code should actually be in main.

**Comment by MathInspector** [comment]
> the StdWrap stuff still doesn't seem to handle flush() correctly.

I consider it a major requirement that it handles all the override stuff perfectly, and is identical to the standard python interpreter.  So tdqm is an excellent way to figure out where the remaining problems are.

`time.sleep(0.2)` is not directly compatible with tkinter, calling this function will completely freeze any tkinter app.  By commenting out the strerr override, it's clear it is working properly because it prints to the console the same as standard python.  https://stackoverflow.com/questions/33175457/time-sleep-equivalent-on-tkinter

The only way to do this imo is to catch calls to `time.sleep` and overwrite them with `self.app.after`.  This is super not safe and hacky, but imo this is necessary.  I already did this with `del` 
https://github.com/MathInspector/MathInspector/blob/dev/mathinspector/console/interpreter.py#L172

The problem is that tkinter has a while loop and has an update function it uses internally.  For the pygame window, I was able to sneak calls to update into the pygame loop (and massively boost performance by being smart about calling update).  There is no way to sneak calls to update into pythons internal implementation of exec tho.

**Comment by casperdcl** [comment]
I don't think this is a `time.sleep` problem:

```python
from tqdm import trange

def foo():
    list(trange(int(1e7)))

foo()
```
```
100%|█████████████████████████████| 10000000/10000000 [00:01<00:00, 6963747.31it/s]
```
hangs for a couple of seconds and prints at the end (incidentally showing only 1sec to complete in the printed bar - so clearly just `mathinspector` taking a v. long time to render).

Increasing the value to `1e8` still doesn't solve the problem: there's just one print and the end:

```
100%|█████████████████████████████| 100000000/100000000 [00:13<00:00, 7363822.13it/s]
```

**Comment by MathInspector** [comment]
> I don't think this is a time.sleep problem:

I am starting to think the same thing

> so clearly just mathinspector taking a v. long time to render

calling time.sleep is probably causing the rendering to slow down because of the way it disrupts the tk root mainloop.
In general, tkinter text has performance problems.  After 2 years I have tamed most of them and know where the soft spots are.  Since you know tqdm so well, you are probably right, and we may both be on the right track.

Right now my guess is we are seeing undefined behavior which results from multiple race conditions in the tk main loop.

**Comment by casperdcl** [comment]
It's not `sleep()` - see my last comment which doesn't use `sleep()` at all (just a large loop)

**Comment by MathInspector** [comment]
I TOTALLY JUST GOT IT!!!!!! =)

File: mathinspector/__init__.py
Comment (MathInspector) [reply mathinspector/__init__.py:112]: > super hacky way to prevent importing other things... The normal way to do this is to move __init__ to app and only import things in __init__ which you want exposed.

I tried this, and when I imported `main` it imported the entire util module and, whenever I imported anything it flooded the result of `dir(mathinspector)` with tons of stuff.  My understanding after researching this was there are 2 schools of thought

- just let everything get imported
- use the super hacky way to do it that used to be "official" way to do it (according to what i read on SO)

Was I doing it wrong, and `dir(mathinspector)` should be clean after `from .app import main`?  Or are you of the first school of thought?  I get that its super hacky, but I felt it was (partially) justifiable.  The negative consequences are unclear to me at this time, curious what could go wrong?

> someone calling mathinspector --help on the CLI does not need the app to run at all
You probably saw this, but it doesn't run the app in that scenario (the 2nd argument in the constructor of the `Help` class is the app, and when this is blank, it only opens the doc window.

> Is there another reason why you do this?
Good question.  I haven't completely made up my mind about `app.py`.  My thinking was:

- I want to minimize the number of files in the root of the mathinsector/ folder
    * I think this helps people who want to read the code and get involved figure things out a lot faster when done right
- the `__init__` files big job is to be responsible for calling `main`, so that function should be defined in that file
- The App class is only used a single time in the entire code base, in main
- Blocks of code which are only used for one thing with other parts of code should be next to each other in the same file unless that file gets too big
- For a short class like App, it's not helping code readability to move it into its own file
    * its actually hurting it, since it's hiding something that you need to know about to understand how __init__ works

You may have noticed I care a lot about code purity, naming conventions, and file/dir structure.  In this case, I am not completely convinced keeping App in it's own file `app` is better or worse than `__init__`  For people just starting to read the code base, it saves them a click and a moment of wondering where App came from.  Small wins like that can add up to big improvements in overall code readability and quality.

> I'm going to close my eyes over the other dark magicke of redefining help(), print(), sanity(), etc..

I have been holding my nose on that lol.  Here is how I feel about global variables.  If you know what you are doing, then they are a tool like everything else.  I am not afraid of using global variables, I use them when they are the right tool for the job.  In this case, it looks like this is a good use a global variables.

Comment (MathInspector) [reply mathinspector/__init__.py:112]: tl;dr Is there a way to get the identical output for `dir(mathinspector)` without doing the super hacky thing?

**Comment by casperdcl** [comment]
P.S. Can you locally fetch & rebase once rather than constantly merging so that I don't have to keep rebasing? :)

File: mathinspector/__init__.py
Comment (MathInspector) [reply mathinspector/__init__.py]: Ok I removed the sys.modules stuff https://github.com/MathInspector/MathInspector/pull/38/commits/8c482ca286daee9e7398ce2b4b0d28b1ab931232

It's probably better to have it count as a module when you do things like `inspect.ismodule(mathinspector)` for one thing.  It's no big deal if someone wants to use undocumented parts of the mathinspector api i guess, no reason to lock it down with the sys.modules hack

File: mathinspector/__main__.py
Comment (casperdcl) [inline mathinspector/__main__.py:65]: this logic should be in the `main()` function itself

**Comment by MathInspector** [comment]
ok all of the doc related things I wanted to do for this PR are done now

File: mathinspector/__main__.py
Comment (MathInspector) [reply mathinspector/__main__.py:65]: I disagree, as I previously explained, I had been planning to use the `__main__` file exactly like this.  Why do you think it should be in `main()`?

Comment (MathInspector) [reply mathinspector/__main__.py:65]: I thought this was an interesting comment about putting the sys.argv logic in main()

If a user has a script `test.py` that has `mathinspector.main()`, and they call this script with argv's
```
$ python test.py --somearg
```
then this will cause the value of `sys.argv` in the `main()` function to get bad data and the app will fail to launch.  Which would be a bug imo.

Comment (casperdcl) [reply mathinspector/__main__.py:65]: there are lots of reasons but it's easiest to ask the other way around, really - why didn't you? I presume you looked at a few established python libraries and so should've seen they put the `sys.argv` parsing login in a function (of the world's top 15 out of ~330k ranked Python packages, 3 implement `__main__.py`, and they all use a function: https://github.com/pytest-dev/pytest/blob/master/src/pytest/__main__.py, https://github.com/tqdm/tqdm/blob/master/tqdm/__main__.py, https://github.com/ipython/ipython/blob/master/IPython/__main__.py).

Comment (casperdcl) [reply mathinspector/__main__.py:65]: er, no, a user calling `mathinspector.main()` intends to forward `sys.argv` to mathinspector. If they intend otherwise they need to use `mathinspector.main(argv=[])` instead. Once again, check other packages for convention :)

Comment (MathInspector) [reply mathinspector/__main__.py:65]: > it's easiest to ask the other way around, really - why didn't you?

There are lots of reasons.  First of all, I wanted people to be able to use
```python
import mathinspector
mathinspector.main()
```
in their own source code files.  I wanted it to be the same `main()` function the app uses for the cli but to accept python style `*args, **kwargs`.  I wanted to keep the logic for parsing argv in __main__ so that in the future that file can grow much larger and be completely dedicated to a robust cli for mathinspector, and to do so in a way that every new addition to the cli is automatically applied to not only the import interface, but also both the in-app GUI and the math inspector source code files themselves.

It just didn't seem like there was anything wrong with using the `__main__.py` file in this way.  I am very happy with it.

I understand that other python packages don't use `__main__` like this, but I have a lot more reasons why I wanted to keep it like this and it's the result of careful consideration.

What I would like to know is: why do 12 out of the top 15 python packages feel it's important to not use `__main__` in the way I did?

Comment (casperdcl) [reply mathinspector/__main__.py:65]: you're talking about a `main-run` pattern - I can push this in a different branch.

Comment (MathInspector) [reply mathinspector/__main__.py:65]: tbh, that bit with how `__main__` parses sys.argv and passes it as `*args, **kwargs` to the `main()` function in `__init__` is something I was super proud of and until you said you wanted to move it, I had been eagerly awaiting your comment on that because I thought it was that good.

I guess I kind of feel strongly that I want to keep it, and also since I like that pattern so much and just made it up, I'd like to understand more about what the top packages consider to be best practices.

Comment (casperdcl) [reply mathinspector/__main__.py:65]: In this case it's not what you wrote, it's just where you wrote it that's bad. See #43.

Comment (MathInspector) [reply mathinspector/__main__.py:65]: I believe an important part of working with other people is I don't always get to have my way.  In this case, it's truly very important to me that you feel good about whatever resolution we come up with.  I also very much want to learn from you more about the higher levels of python.

I have been programming for my entire life, and have come to the strong conclusion that having a dedicated file in the root of the most important folder in a project that does nothing but import a function from another file and call it once is a bad practice.

I disagree that where I put it was bad.  I claim that the so-called "best practice" of having a blank `__main__` file is in fact misusing the original design of python and why there is a `__main__` file.  It's a way of trying to work against python imo.

Comment (MathInspector) [reply mathinspector/__main__.py:65]: > If they intend otherwise they need to use mathinspector.main(argv=[]) instead. 

I strongly disagree with the concept that `mathinspector.main(argv=[])` is a preferable interface to `mathinspector.main(*args, **kwargs)` for calling a function named `main` after doing `import mathinspector` in either a file or from the standard python interpreter.

> Once again, check other packages for convention :)

You keep on making references to authority, and with all due respect to your very high skill level at programming; I assert that I am an expert on the philosophy of writing good code with literally a lifetime of experience across over 10 programming languages, over 30 years, and I have been the CTO of multiple successful companies whose technological infrastructure I created from scratch and personally maintained while working with a team of people who reported to me.

I have asked you multiple times to justify your position because I want to learn from you, because I value and respect your opinion.  So "once again", please justify your statement that the "where I put it was bad".

Comment (casperdcl) [reply mathinspector/__main__.py:65]: Ah sorry but I wasn't offering an *explanation* but a *reason* ("everyone does it so you should too" - if viewed as an *explanation* this would of course classify as a logical fallacy).

I could spend about half an hour writing a comprehensive explanation which would be the beginnings of a subsection in a book on Python which I admittedly should write at some point in the future. However for now, to be blunt if I'm not being paid to do something then it's a hobby. And if it's a hobby I don't owe anyone an explanation. However I did anticipate a reasonable view that *reason sans explanation* does not warrant inclusion in `master` and thus - rather than commit here - opened a separate PR which you are free to ignore.

I might find time to respond to some of your review comments in #43 but I'm afraid unlikely to delve into comparisons of design patterns and APIs.

Comment (MathInspector) [reply mathinspector/__main__.py:65]: I would like to merge this PR into master as is, and close #43 in a few days unless you reply to my review comments.  Does that sound good to you?

It's not necessary to reply to any of those review comments.  If you do, I would be very interested in having a philosophical discussion about the specific technical details of what's going on here.

> to be blunt if I'm not being paid to do something then it's a hobby. And if it's a hobby I don't owe anyone an explanation

Well said.  I am also not being paid for this, although it's more than just a hobby for me, since I am so personally invested in this project at this point.  If anything, I feel like I am in your debt for your amazing contribution to the python package structure.  It seems fair to heavily weight your opinion when making decisions moving forward for this reason.

If you claim something is True, that's generally good enough for me, and I don't need an explanation, nor think you owe me one.  In this case, we are talking about something I had been planning for 6 months and have carefully considered and researched.  I hope you have enjoyed working on this project, and that whatever reasons you have in the first place for being the first contributor, that we can build on those with this project so it's not just mine anymore, but belongs to the entire community.

> I could spend about half an hour writing a comprehensive explanation which would be the beginnings of a subsection in a book on Python which I admittedly should write at some point in the future

I was asking for an explanation because I would like to read that book!

> I did anticipate a reasonable view that reason sans explanation does not warrant inclusion in master

Thanks for this.

**Pull request merged**
